### ISSUE TITLE
Reduction pass idea: loop construct to selection construct

### ISSUE THREAD

**Issue by afd** [KhronosGroup/SPIRV-Tools#2122]
I want to implement a spirv-reduce reduction pass for structured control flow that basically replaces a loop with its body, along the lines of:

while(c) { B; } -> if(c) { B; }

I propose doing this as follows:

Suppose we have a loop header block of the form:

```
OpLabel LH
(OpPhi instructions)
...
OpLoopMerge MB CT
(OpBranch B) or (OpBranchConditional X B1 B2)
```

Suppose further that LH strictly dominates MB (so that if LH is reachable, so is MB).  The potential need for these is discussed below.

Under these conditions we can turn the loop construct into a selection construct as follows:

(1) Remove the entire construct headed at CT from the module.

(2) Change the LH block to:

```
OpLabel LH
...
(transformed OpPhi instructions)
...
OpSelectionMerge MB
(OpBranchConditional <some boolean> B MB) or (OpBranchConditional X B1 B2)
```

where:
	- OpPhi instructions have values previously flowing form the loop back edge removed
	- If LH ended with "OpBranch B", it now ends with an OpConditional that can branch to B or MB (the condition doesn't especially matter; true might work well).

(3) Any edge A -> CT or A -> MB, where block A is part of the loop construct headed at LH, is replaced with an edge A -> MB' where MB' is the merge block associated with the inner-most structured control-flow construct that contains A.  We will have MB' = MB if A is not nested in any further structured control-flow construct.

The restriction is needed because after the transformation MB will be reachable and LH will be required to strictly dominate it.  If MB were previously unreachable then it might appear at an inconvenient place in the module.

I also thought we would need this restriction:

- Restriction 2: MB strictly post-dominates LH (so that we don't have returns inside the loop for example)

...but actually I don't think this is needed.  A return from inside the resulting selection construct would be just as legitimate as the original return from the loop construct would have been.

**Comment by alan-baker** [comment]
You may run afoul of the structured control flow rules in the following way. From anywhere in the most deeply nested loop you can branch directly to the loop's merge or continue target. Once this is transformed into a selection those branches may become illegal. You can only branch to the deepest selection construct's merge. So if the break or continue is within a selection construct you will not be able to transform the code as described.

**Comment by afd** [comment]
Hi Alan

I think I have catered for this in the above in case (3): if we have an edge A -> CT or A -> MB branching to a loop's continue target or merge block then we change this to A -> MB' (note the prime), where MB' is the merge block for the selection construct most tightly containing A.  So if we had the structured control flow equivalent of:

while(...) { // 1
  if (...) { // 2
     if(...) { // 3
        continue;
     }
  }
}

the continue would get changed to a branch that jumps to the merge block for the if statement marked as (3).

Does that make sense?

**Comment by alan-baker** [comment]
Yes, that makes sense. I think I misread it the first time through.

**Comment by afd** [comment]
Closed by 6679d5df89834c58ec3357fd5fcb1e5e554ad543 / #2164
